File: WolfDen/Game/Behaviours/CombatBehaviour/RangedCombatBehaviourTemplate.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using WolfDen.Game;
using WolfDen.Game.Interfaces;

namespace WolfDen.Game.Behaviours.CombatBehaviour
{
    public class RangedCombatBehaviourTemplate : NormalCombatBehaviourTemplate
    {
        public int MinRange { get; set; }
        public int MaxRange { get; set; }

        public override void AddBehaviour(Unit unit)
        {
            var moveBehaviour = unit.Behaviours.Values.OfType<IMoveBehaviour>().FirstOrDefault();
            if (moveBehaviour == null)
                throw new Exception("No movebehaviour found while generating combat behaviour. Make sure movebehaviour is added first.");
            var combatBehaviour = new RangedCombatBehaviour(unit.Game, unit, MaxHealth, UnitType, moveBehaviour, DamageChart, MinRange, MaxRange);
            combatBehaviour.Id = unit.Id + Class;
            unit.Behaviours.Add(combatBehaviour.Id, combatBehaviour);
        }
    }
}

File: WolfDen/wwwroot/ts/Behaviours/CombatBehaviour/RangedCombatBehaviourTemplate.ts
/// <reference path="../../Unit.ts" />

namespace TurnBasedBoardGame {
    export class RangedCombatBehaviourTemplate extends NormalCombatBehaviourTemplate {
        MinRange: number;
        MaxRange: number;

        AddBehaviour(unit: Unit): void {
            let self = this;
            let combatBehaviour = new RangedCombatBehaviour();
            Object.assign(combatBehaviour, self);
            let moveBehaviour = null;
            for (let key in unit.Behaviours)
                if (IsIMoveBehaviour(unit.Behaviours[key]))
                    moveBehaviour = unit.Behaviours[key];
            combatBehaviour.MoveBehaviour = moveBehaviour;
            combatBehaviour.Unit = unit;
            combatBehaviour.Game = unit.Game;
            combatBehaviour.CurrentHealth = self.MaxHealth;
            combatBehaviour.Id = unit.Id + self.Class;
            combatBehaviour.Initialize();
            unit.Behaviours[combatBehaviour.Id] = combatBehaviour;
        }
    }
}